docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+    In other words, record fields or columns that are never accessed never need to be typed.
+
+    Additionally if we assume transformations are given correct types, type inference can determine the columns and column types that the user expects from the database.
+    Storage can then implicitly filter out any types that do not match the user's expectation.
+
+### Typing flux functions using lists of records
+
+The goal of this 2nd iteration is **not** to assign an accurate type to every Flux transformation.
+As it stands, it is not possible to accurately describe certain Flux transformations with a type signature.
+Instead we should give every transformation a type signature of the form:
+
+```
+forall ['r, 's] where 'r:Rec, 's:Rec (tables: ['r], ...) -> ['s]
+```
+
+In words, every Flux transformation is a function which takes in a list of record kinds and returns a list of record kinds.
+This gives us parity with the current system while allowing us to iteratively assign more accurate and descriptive types, giving us the ability to catch more errors at compile time.
+
+Below I attempt to assign accurate types to common Flux transformations.
+
+* from
+
+    ```
+    forall ['r, 's] where 's: StrRec (org: str, bucket: str) -> [ {_time: time | _start: time | _stop: time | _value: 'r | 's} ]
+    ```
+    This polymorphic type signature says that `from` is a function that returns a list of records.
+    These records have fields `_time`, `_start`, `_stop`, and `_value`.
+    The `_value` field can be of any type.
+    All other fields represented by the row variable `'s` must be of type _string_.
+    Note the last part corresponds to the tag values of a series.
+
+    An example of a program that will fail to compile is:
+    ```
+    from(bucket: "telegraf")
+        |> range(start: -1h)
+        |> filter(fn: (r) => r.host == 1) // error: {host: int | 't} is not of kind StrRec
+    ```
+    The reason for failed compilation will be clear soon, but basically `range` and `filter` are just pass through transformations.
+    They don't extend or modify the types of their inputs.
+    As a result `filter` expects to receive the same list type that `from` returns and so it expects `r.host` to be of type _string_.
+    This contradicts the usage of `r.host` as an _int_.
+
+* range
+
+    ```
+    forall ['r] where 'r:Rec (start: time, stop: time, tables: [ {_start: time | _stop: time | 'r} ]) -> [ {_start: time | _stop: time | 'r} ]
+    ```
+    Let's focus on the `tables` parameter.
+    `range` takes a `tables` parameter that is a list of records that must have fields `_start` and `_stop` and returns the same list type.
+
+* filter
+
+    ```
+    forall ['r] where 'r:Rec (fn: (r: 'r) -> bool, tables: ['r]) -> ['r]
+    ```
+    `filter` takes a list of records and returns that same list of records.
+
+    From the type signatures of `range` and `filter` it should be clear why the above example program fails unification.
```

Comment:
Will the usage information we can infer from the filter predicate function propagate both up and down the chain?


For example if the user does this `from() |> filter(fn: (r) => r.host == "a")` since `'r` is unmodified in `filter` then we infer that the record produced by `from` must have a `host` of type string?

Reply:
Yes, Hindley-Milner ensures this.